### ISSUE TITLE
Evaluate `stop_` calls in `map_` environments

### ISSUE THREAD

**Issue by lionel-** [tidyverse/purrr#837]
To improve rlang backtraces. See https://github.com/r-lib/rlang/issues/1023

**Comment by hadley** [comment]
```R
a <- function(x) b()
b <- function() c()
c <- function() rlang::abort("error")

f <- function() g()
g <- function() h()
h <- function() purrr::map(1:3, a)

rlang::catch_cnd(h())
```

```R
Backtrace:
    ▆
 1. └─global h()
 2.   └─purrr::map(1:3, a)
 3.     └─global .f(.x[[i]], ...)
 4.       └─global b()
 5.         └─global c()
 ```

**Comment by hadley** [comment]
Attempt to create something a little closer to the original issue in rlang:

```R
library(tidyverse)

a <- function(x) b()
b <- function() c()
c <- function() "x"

f <- function(x) g(x)
g <- function(x) h(x)
h <- function(x) purrr::map_int(x, a)

df <- data.frame(x = 1:10)

df %>% 
  filter(x < 5) %>% 
  filter(TRUE) %>% 
  mutate(f(x))
```

With a full backtrace I see:

```
<error/dplyr:::mutate_error>
Error in `mutate()`:
! Problem while computing `..1 = f(x)`.
Caused by error:
! Can't coerce element 1 from a character to a integer
---
Backtrace:
     ▆
  1. ├─df %>% filter(x < 5) %>% filter(TRUE) %>% mutate(f(x))
  2. ├─dplyr::mutate(., f(x))
  3. ├─dplyr:::mutate.data.frame(., f(x))
  4. │ └─dplyr:::mutate_cols(.data, dplyr_quosures(...), caller_env = caller_env())
  5. │   ├─base::withCallingHandlers(...)
  6. │   └─mask$eval_all_mutate(quo)
  7. │     └─dplyr (local) eval()
  8. └─global f(x)
  9.   └─global g(x)
 10.     └─global h(x)
 11.       └─purrr::map_int(x, a)
```

Which looks ok to me? If there is a problem I'd say it's coming from `mutate()`, not from purrr.

**Comment by lionel-** [comment]
yup it looks like I was confused in the original issue. But we should still evaluate in the current frame as a rule to avoid discontinuities:

```r
f <- function(x) g(x)
g <- function(x) h(x)
h <- function(x) purrr::map_int(x, \(x) 1:2)
f(1)
```

```
Error in `stop_bad_type()` at lowliner/R/conditions.R:201:2:
! Result 1 must be a single integer, not an integer vector of length 2
---
Backtrace:
    ▆
 1. ├─global f(1)
 2. │ └─global g(x)
 3. │   └─global h(x)
 4. │     └─purrr::map_int(x, function(x) 1:2)
 5. └─purrr:::stop_bad_element_vector(...)
 6.   └─purrr:::stop_bad_vector(...) at lowliner/R/conditions.R:226:2
 7.     └─purrr:::stop_bad_type(x, expected, actual = actual, what = what, arg = arg, recycle = recycle,
  message = message, .subclass = c(.subclass,
    "purrr_error_bad_vector")) at lowliner/R/conditions.R:201:2
```

Though it's more important to pass the correct call so that this part of the backtrace is hidden by default. The rlang C library provides tools for both tasks.

**Comment by hadley** [comment]
Ah ok. In that case it's already fixed as part of #930:

```
   ▆
 1. └─global f(1)
 2.   └─global g(x)
 3.     └─global h(x)
 4.       └─purrr::map_int(x, function(x) 1:2)
 5.         └─purrr:::stop_bad_element_length(...)
 6.           └─cli::cli_abort(...) at purrr/R/conditions.R:73:2
 7.             └─rlang::abort(...)
```
